### PR TITLE
Move TSCH neighbors to a neighbor table

### LINKED ISSUE #1054: Neighbor table consistency
There are a number of modules in Contiki-NG that use neighbor tables:
- RPL candidate parents
- link stats
- IPv6 neighbors (called `ds6_neighbors`)
- TSCH neighbors (not an actual `NBR_TABLE`, but used in a similar role)
- other tables (to be reviewed!)

**The problem:** There is no attempt to ensure that the tables contain the same entries.

This can (and does) lead to issues. One is that the TSCH neighbors and link stats are removed much more aggressively than RPL candidate parents.
- If there is an entry for RPL parent, but a TSCH neighbor cannot be added for this parent, the parent cannot be probed.
- If the entry for link stats is deleted for a RPL parent, it can be re-added back later with a better metric. This causes the routing tree convergence time to become very long.

**Proposed solution**: introduce a global consistency mechanism between the tables.

Details:
- Could use an approach with bitmaps to implement that: only remove a neighbor when all of TSCH, RPL and link stats have freed that neighbor, i.e. the bitmap is zero.
- Common neighbor replacement policy can now be implemented, e.g. the "least recently used" (LRU) policy.
- Additionally, there should be a locking / prioritization option (the existing one may be sufficient?). For example, a TSCH timesource or RPL preferred parent should never be removed to free space for a new neighbor that is neither one of these.
- Allow to disable the IPv6 neighbor table completely, to save RAM. As discussed with George, in the normal Contiki-NG use case there is no need to keep that table at all, because the MAC address / link address mapping can be done on the fly.

### PR DESCRIPTION
**Pull request by atiselsts**
Reduces memory usage and partially fixes #1054 .

### CODE REVIEW

**Comment by atiselsts** [comment]
@yatch Some of the 6top tests are failing. I think because `13-ieee802154/code-6tisch/project-conf.h` configures `NBR_TABLE_CONF_MAX_NEIGHBORS         2` and relies on the neighbor table having exaclty 2 free spaces. Now that TSCH is also using neighbor table this is not always true.

**Comment by yatch** [comment]
@atiselsts I will handle the failures.

**Comment by yatch** [comment]
@atiselsts I found `tsch_queue_init()` now consumes two `nbr` entries: one for `tsch_eb_address`, and the other for `tsch_broadcast_address`. Increased `NBR_TABLE_CONF_MAX_NEIGHBORS` to `4` and modified the code a little bit. Please find attached a file which is a patch resoloving the failures. Tested on my machine locally.

[pr-1239.txt](https://github.com/contiki-ng/contiki-ng/files/4510867/pr-1239.txt)

File: os/net/mac/tsch/tsch-types.h [hunk 1]
```diff
@@ -107,9 +107,7 @@ struct tsch_packet {
 
 /** \brief TSCH neighbor information */
 struct tsch_neighbor {
-  /* Neighbors are stored as a list: "next" must be the first field */
-  struct tsch_neighbor *next;
-  linkaddr_t addr; /* MAC address of the neighbor */
+//  linkaddr_t addr; /* MAC address of the neighbor */
```
Comment (yatch) [inline os/net/mac/tsch/tsch-types.h:110]: I guess you're aware of this; but just in case...

**Comment by atiselsts** [comment]
Now ready for review.

File: os/net/mac/tsch/tsch-queue.c [hunk 2]
```diff
@@ -211,16 +213,14 @@ tsch_queue_remove_nbr(struct tsch_neighbor *n)
...
-
       tsch_release_lock();
 
       /* Flush queue */
       tsch_queue_flush_nbr_queue(n);
 
-      /* Free neighbor */
-      memb_free(&neighbor_memb, n);
+      /* Unlock and free neighbor */
+      nbr_table_unlock(tsch_neighbors, n);
```
Comment (simonduq) [inline os/net/mac/tsch/tsch-queue.c:222]: No need to unlock really, remove will do the job ("Locked" only means "not GC-able")

Comment (atiselsts) [reply os/net/mac/tsch/tsch-queue.c]: Thanks, updated the code.

**Comment by atiselsts** [comment]
> How tested is this?

I checked that adding and removing neighbors work fine.

The regression tests pass, so apparently the new approach of using the nbr-table API to get neighbor's address works as well.

**Pull request merged**